rflx/pyrflx/message.py

```diff
@@ -0,0 +1,234 @@
+            return Number(typeval.size)
+        raise RuntimeError(f"cannot determine length of {fld}")
+
+    def _has_length(self, fld: str) -> bool:
+        prv = self._prev_field(fld)
+        for l in self._model.outgoing(model.Field(prv)):
+            if isinstance(self.__simplified(l.length), Number):
+                return True
+        return isinstance(self._fields[fld].typeval, ScalarValue)
+
+    def _get_first(self, fld: str) -> Number:
+        for l in self._model.incoming(model.Field(fld)):
+            if self.__simplified(l.condition) == TRUE and l.first != UNDEFINED:
+                first = self.__simplified(l.first)
+                break
+        else:
+            prv = self._prev_field(fld)
+            first = self.__simplified(Add(self._fields[prv].first, self._fields[prv].length))
+        if isinstance(first, Number):
+            return first
+        raise RuntimeError(f"cannot determine first of {fld}")
+
+    def _has_first(self, fld: str) -> bool:
+        prv = self._prev_field(fld)
+        first = self.__simplified(Add(self._fields[prv].first, self._fields[prv].length))
+        return isinstance(first, Number)
+
+    def set(self, fld: str, value: Any) -> None:
+        if not isinstance(value, self._fields[fld].typeval.accepted_type):
+            raise TypeError(
+                f"cannot assign different types: {self._fields[fld].typeval.accepted_type.__name__}"
+                f" != {type(value).__name__}"
+            )
+        try:
+            field = self._fields[fld]
+            field.first = self._get_first(fld)
+            field.length = self._get_length(fld)
+            self._fields[fld].typeval.assign(value, True)
+            if all(
+                [
+                    self.__simplified(o.condition) == FALSE
+                    for o in self._model.outgoing(model.Field(fld))
+                ]
+            ):
+                self._fields[fld].typeval.clear()
+                raise ValueError("value does not allow path to next field")
+        except KeyError:
+            raise KeyError(f"cannot access field {fld}")
+        if isinstance(field.typeval, OpaqueValue) and field.typeval.length != field.length.value:
+            flength = field.typeval.length
+            field.typeval.clear()
+            raise ValueError(f"invalid data length: {field.length.value} != {flength}")
+        self._preset_fields(fld)
+
+    def _preset_fields(self, fld: str) -> None:
+        nxt = self._next_field(fld)
+        while nxt and nxt != model.FINAL.name:
+            field = self._fields[nxt]
+            try:
+                field.first = self._get_first(nxt)
+                field.length = self._get_length(nxt)
+                if (
+                    field.set
+                    and isinstance(field.typeval, OpaqueValue)
+                    and field.typeval.length != field.length.value
+                ):
+                    raise RuntimeError
+            except RuntimeError:
```

Comment:
Catching `RuntimeError` could hide internal Python errors (like `RecursionError`).

Reply:
I think you shouldn't use exceptions in this case at all.